### PR TITLE
Torch Function modes x torch.compile tutorial

### PR DESCRIPTION
**Pull request by mlazos**
Fixes #ISSUE_NUMBER

## Checklist

- [ ] The issue that is being fixed is referred in the description (see above "Fixes #ISSUE_NUMBER")
- [ ] Only one issue is addressed in this pull request
- [ ] Labels from the issue that this PR is fixing are added to this pull request
- [ ] No unnecessary issues are included into this pull request.

### CODE REVIEW

File: recipes_source/torch_compile_torch_function_modes.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
+"""
+(beta) Utilizing Torch Function modes with torch.compile
+============================================================
+
+**Author:** `Michael Lazos <https://github.com/mlazos>`_
+"""
+
+#########################################################
+#  This tutorial covers how to use a key torch extensibility point,
```
Comment (svekars) [inline recipes_source/torch_compile_torch_function_modes.py:9]:
```suggestion
#  This recipe covers how to use a key torch extensibility point, 
```

File: recipes_source/torch_compile_torch_function_modes.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
+"""
+(beta) Utilizing Torch Function modes with torch.compile
+============================================================
+
+**Author:** `Michael Lazos <https://github.com/mlazos>`_
+"""
+
+#########################################################
+#  This tutorial covers how to use a key torch extensibility point, 
+#  torch function modes, in tandem with torch.compile to override
```
Comment (svekars) [inline recipes_source/torch_compile_torch_function_modes.py:10]:
```suggestion
#  torch function modes, in tandem with ``torch.compile`` to override 
```

File: recipes_source/torch_compile_torch_function_modes.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+(beta) Utilizing Torch Function modes with torch.compile
+============================================================
+
+**Author:** `Michael Lazos <https://github.com/mlazos>`_
+"""
+
+#########################################################
+#  This tutorial covers how to use a key torch extensibility point, 
+#  torch function modes, in tandem with torch.compile to override 
+#  the behavior of torch ops at trace time, with no runtime overhead.
```
Comment (svekars) [inline recipes_source/torch_compile_torch_function_modes.py:11]: ops => operators?
```suggestion
#  the behavior of torch operators, also know as **ops**, at trace time, with no runtime overhead.
```

File: recipes_source/torch_compile_torch_function_modes.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+#
+# .. note::
+#
+#    This tutorial requires PyTorch 2.7.0 or later.
```
Comment (svekars) [inline recipes_source/torch_compile_torch_function_modes.py:15]:
```suggestion
#    This recipe requires PyTorch 2.7.0 or later.
```

File: recipes_source/torch_compile_torch_function_modes.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+
+#####################################################################
+# Rewriting a torch op (torch.add -> torch.mul)
+# ~~~~~~~~~~~~~~~~~~~~~
```
Comment (svekars) [inline recipes_source/torch_compile_torch_function_modes.py:20]:
```suggestion
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

File: recipes_source/torch_compile_torch_function_modes.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+#  torch function modes, in tandem with torch.compile to override 
+#  the behavior of torch ops at trace time, with no runtime overhead.
+#
+# .. note::
+#
+#    This tutorial requires PyTorch 2.7.0 or later.
+
+
+#####################################################################
+# Rewriting a torch op (torch.add -> torch.mul)
```
Comment (svekars) [inline recipes_source/torch_compile_torch_function_modes.py:19]:
```suggestion
# Rewriting a torch op (``torch.add`` -> ``torch.mul``)
```

File: recipes_source/torch_compile_torch_function_modes.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+# exit cleanly if we are on a device that doesn't support ``torch.compile``
+if torch.cuda.get_device_capability() < (7, 0):
+    print("Exiting because torch.compile is not supported on this device.")
+    import sys
+    sys.exit(0)
+
+from torch.overrides import BaseTorchFunctionMode
+
+# Define our mode, Note: BaseTorchFunctionMode
```
Comment (svekars) [inline recipes_source/torch_compile_torch_function_modes.py:35]:
```suggestion
# Define our mode, Note: ``BaseTorchFunctionMode``
```

File: recipes_source/torch_compile_torch_function_modes.py [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+x = torch.rand(2, 2)
+y = torch.rand_like(x)
+
+with AddToMultiplyMode():
+    z = test_fn(x, y)
+
+assert torch.allclose(z, x * y * x)
+
+# The mode can also be used within the compiled region as well like so
```
Comment (svekars) [inline recipes_source/torch_compile_torch_function_modes.py:56]:
```suggestion
# The mode can also be used within the compiled region as well like this:
```

File: recipes_source/torch_compile_torch_function_modes.py [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+x = torch.rand(2, 2)
+y = torch.rand_like(x)
+z = test_fn(x, y)
+
+assert torch.allclose(z, x * y * x)
+
+######################################################################
+# Conclusion
+# ~~~~~~~~~~
+# In this tutorial we demonstrated how to override the behavior of torch.* operators
```
Comment (svekars) [inline recipes_source/torch_compile_torch_function_modes.py:72]:
```suggestion
# In this recipe we demonstrated how to override the behavior of ``torch.*`` operators
```

File: recipes_source/torch_compile_torch_function_modes.py [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+# using torch function modes from within torch.compile. This enables users to utilize
```
Comment (svekars) [inline recipes_source/torch_compile_torch_function_modes.py:73]:
```suggestion
# using torch function modes from within ``torch.compile``. This enables users to utilize
```

File: recipes_source/torch_compile_torch_function_modes.py [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+# the extensibility benefits of torch function modes without the runtime overhead
+# of calling torch function on every op invocation. 
+# 
+# * `Extending Torch API with Modes <https://pytorch.org/docs/stable/notes/extending.html#extending-all-torch-api-with-modes>`__ - Other examples and backgroun on Torch Function modes.
```
Comment (svekars) [inline recipes_source/torch_compile_torch_function_modes.py:77]:
```suggestion
# * See `Extending Torch API with Modes <https://pytorch.org/docs/stable/notes/extending.html#extending-all-torch-api-with-modes>`__  for other examples and background on Torch Function modes.
```

**Review by svekars — COMMENTED** [review]
Looks good, added some minor suggestions.

**Review by williamwen42 — COMMENTED** [review]
Content lgtm

**Comment by svekars** [comment]
it seems to be failing: https://github.com/pytorch/tutorials/actions/runs/14373299892/job/40300235591?pr=3320#step:9:4837
Can you try to put on  a linux.g5.4xlarge.nvidia.gpu in this file: https://github.com/pytorch/tutorials/blob/main/.jenkins/metadata.json

File: .jenkins/metadata.json [hunk 12]
```diff
@@ -49,6 +49,9 @@
   "recipes_source/compiling_optimizer_lr_scheduler.py": {
      "needs": "linux.g5.4xlarge.nvidia.gpu"
   },
+  "recipes_source/torch_compile_torch_function_modes.py" {
+    "needs": "linux.g5.4xlarge.nvidia.gpu"
+  },
```
Comment (svekars) [inline .jenkins/metadata.json:54]:
```suggestion
```
